[New Model]: SpecialCharacteristicMeasurement v.1.0.0

LuisRickert opened the pull request:
## Description
<!-- Please provide a short description about what this PR changes and reference an issue that was initially created to introduce the new aspect model -->

 -->
This PR introduces a data model for special characteristics measurements. The initial model was presented with this issue #781. An example can be found below:

```json 
{
  "localIdentifiers": [
    {
      "value": "SN12345678",
      "key": "partInstanceId"
    }
  ],
  "customerPartID": "BPN0000000",
  "bpnlProperty": "BPNL0123456789ZZ",
  "revisionIndex": "01",
  "measurementType": "eOMtThyhVNLWUZNRcBaQKxI",
  "characteristicID": "S1",
  "results": [
    {}
  ]
}
```

This data model should be included in future into the Traceability KIT [CX-0125](https://catenax-ev.github.io/docs/standards/CX-0125-TraceabilityUseCase)

This PR replaces the old [PR-783](https://github.com/eclipse-tractusx/sldt-semantic-models/pull/783) since it has generated artifacts in it. 

Closes #781 

<!-- The MS2 and MS3 criteria are intended for merges to the main-branch. For small bug-fixes or during the model development, for instance, when merging to a feature branch, you may decide to not fill out the checklists. However, we recommend to follow the MS2 checklist during the development. The MS3 checklist becomes relevant for merges to the main-branch. -->
## MS2 Criteria
(to be filled out by PR reviewer)
- [x] the model **validates** with the SAMM SDS SDK in the version specified in the Readme.md of this repository by the time of the MS2 check  (e.g., 'java -jar samm-cli.jar aspect \<path-to-aspect-model\> validate ). The  SAMM CLI is available [here](https://eclipse-esmf.github.io/esmf-developer-guide/tooling-guide/samm-cli.html) and in [GitHub](https://github.com/eclipse-esmf/esmf-sdk/releases/tag/v2.9.5)
- [x] use **Camel-Case** (e.g., "MyModelElement" or "TimeDifferenceGmtId", when in doubt follow https://google.github.io/styleguide/javaguide.html#s5.3-camel-case)
- [x] the identifiers for all model elements **start with a capital letter** except for properties
- [x] the identifier for **properties starts with a small letter**
- [x] all model elements **at least contain the fields "preferred name" and "description"** in English language. The description must be comprehensible. It is not required to write full sentences but style should be consistent over the whole model
- [x] Property and the referenced Characteristic should not have the same name
- [x] the versioning in the URN **follows semantic versioning**, where minor version bumps are backwards compatible and major version bumps are not backwards compatible. 
- [x] use **abbreviations only when necessary** and if these are sufficiently common
- [x] **avoid redundant prefixes in property names** (consider adding properties to an enclosing Entity or even adapt the namespace of the model elements, e.g., instead of having two properties `DismantlerId` and `DismantlerName` use an Entity `Dismantler` with the properties `name` and `id` or use a URN like `io.catenax.dismantler:0.0.1`)
- [x] fields `preferredName` and `description` are not the same
- [x] **`preferredName` should be human readable** and follow normal orthography (e.g., no camel case but normal word separation)
- [x] name of aspect is singular except if it only has one property which is a Collection, List or Set. In theses cases, the aspect name is plural.
- [x] units are referenced from the SAMM unit catalog whenever possible
- [x] **use constraints** to make known constraints from the use case explicit in the aspect model 
- [x] when relying on **external standards**, they are referenced through a **"see"** element
- [x] all properties with an [simple type](https://eclipse-esmf.github.io/samm-specification/2.1.0/datatypes.html) have an example value
- [x] metadata.json exists with status "release"
- [x] generated json schema validates against example json payload
- [x] file RELEASE_NOTES.md exists and contains entries for proposed model changes 
- [x] all contributors to this model are mentioned in copyright header of model file

## MS3 Criteria
(to be filled out by semantic modeling team before merge to main-branch)
- [x] All required reviewers have approved this PR (see reviewers section)
- [x] The new aspect (version) will be implemented by at least one data provider
- [x] The new aspect (version) will be consumed by at least one data consumer
- [x] There exists valid test data
- [x] In case of a new (incompatible) major version to an existing version, a migration strategy has been developed
- [x] The model has at least version '1.0.0'
- [x] If a previous model exists, model deprecation has been checked for previous model
- [x] The release date in the Release Note is set to the date of the MS3 approval

LuisRickert commented:
> ### Validation Report for io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl
> #### Semantic violations were found:
> ```turtle
> > :Uuid :
>   ---> Error at line 27 column 5
>    | 
> 27 |      samm:properties (:uuidV4Property ) ; ( ) ; ( :bpnaProperty ) ; . . ; ; . a  ; .
>    |      ^^^^^^^^^^^^^^^ Property samm:properties is used 2 times on :Uuid, but may only be used 1 time.
>    | 
> 
>   ---> Error at line 27 column 5
>    | 
> 27 |      samm:properties (:uuidV4Property ) ; ( ) ; ( :bpnaProperty ) ; . . ; ; . a  ; .
>    |      ^^^^^^^^^^^^^^^ Property samm:properties is used 2 times on :Uuid, but may only be used 1 time.
>    | 
> ```

It appears that this is caused by the data model Serial Part. If the validation is run locally on the SerialPart the same error is returned.

johannsvarela commented on io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,63 @@
+
```
Please add the license information.

```suggestion
#######################################################################
# Copyright(c) 2024 BASF SE
# Copyright(c) 2024 Bayerische Motoren Werke Aktiengesellschaft(BMW AG)
# Copyright(c) 2024 Fraunhofer-Gesellschaft zur Foerderung der angewandten Forschung e.V.(represented by Fraunhofer ISST & Fraunhofer IML)
# Copyright(c) 2024 German Edge Cloud GmbH & Co. KG
# Copyright(c) 2024 Henkel AG & Co. KGaA
# Copyright(c) 2024 Mercedes Benz AG
# Copyright(c) 2024 Robert Bosch Manufacturing Solutions GmbH
# Copyright(c) 2024 SAP SE
# Copyright(c) 2024 Siemens AG
# Copyright(c) 2024 T-Systems International GmbH
# Copyright(c) 2024 ZF Friedrichshafen AG
# Copyright(c) 2024 Contributors to the Eclipse Foundation
#
# See the NOTICE file(s) distributed with this work for additional information regarding copyright ownership.
#
# This work is made available under the terms of the Creative Commons Attribution 4.0 International(CC-BY-4.0) license, which is available at https://creativecommons.org/licenses/by/4.0/legalcode.
#
# SPDX-License-Identifier: CC-BY-4.0
#######################################################################
```

LuisRickert replied:
Done

johannsvarela commented on io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,63 @@
+@prefix samm-u: <urn:samm:org.eclipse.esmf.samm:unit:2.1.0#> .
+@prefix serialPart: <urn:samm:io.catenax.serial_part:3.0.0#> .
+
+:Measurement a samm:Aspect ;
```
Is there a reason why you name the aspect model _'Measurement'_? If you want to create an aspect model for special characteristics does it not make more sense to name the aspect SpecialCharacteristic or SpecialCharacteristicMeasurment?

LuisRickert replied:
No, the idea was to show that it is related to specialCharacteristics through the path in the name.

**io.catenax.specialCharacteristics.measurements**/1.0.0/

I guess this should not be a serious problem in renaming the model

LuisRickert replied:
renamed it to specialCharacteristicMeasurement

johannsvarela commented on io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,63 @@
+   samm:operations ( ) ;
+   samm:events ( ) .
+
+:customerPartId a samm:Property ;
```
Since the _localIdentifier_ was referenced from the SerialPart i would also suggest to use the reference _customerPartId_ from the SerialPart instead of define a new property with the name customerPartId.

LuisRickert replied:
good point, will discuss this with Jochen and Sebastian. If we move the attribute into local identifier then we should explicitly reference them in the KIT

johannsvarela replied:
@LuisRickert the customerPartId is **not part** of the localIdentifier, therefore **don't move this attribute into the localIdentifier** array. What i mean is to integrate the existing customerPartId attribute from the SerialPart into this aspect model. 

You've already done this with the localIdentifier:
![image](https://github.com/user-attachments/assets/058c05e2-7768-4cb0-bd2e-97c04413bcac)

### Suggestion
You should do the same with the customerPartId doing this steps:
1. **Remove** the **":customerPartId" property**
![image](https://github.com/user-attachments/assets/b196df57-60b5-42af-9e7f-6848154d6c06)

2. **Change** the defined property entry **":customerPartId"** at the Measurement aspect
![image](https://github.com/user-attachments/assets/06777744-ff24-4b4b-a39c-aa845e5e6605)

3. **to "serialPart:customerPartId"**
  ![image](https://github.com/user-attachments/assets/3af217ed-e9ac-43f5-96eb-a9778736bcbc)

4. Then the integration of the customerPartId from SerialPart is done! 😄

LuisRickert replied:
ahh got it, i just had a missunderstanding :)

LuisRickert replied:
done

johannsvarela commented on io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,63 @@
+:Measurement a samm:Aspect ;
+   samm:preferredName "SpecialCharacteristicMeasurment"@en ;
+   samm:description "Contains the Measurment for a special characteristic"@en ;
+   samm:properties ( serialPart:localIdentifiers bpn:bpnlProperty :customerPartId :characteristicId :results [ samm:property :measurementType; samm:optional true ] :revisionIndex ) ;
```
Is the **bpnlProperty** really needed? The same information is stored in the _manufacturerID_ of the _LocalIdentifier_? This information would therefore be redundant.

Unless you don't mean the manufactuererId with the bpnlProperty? If that is the case, then I would still suggest using a unique name for this property to know exactly what is meant.

LuisRickert replied:
> good point, will discuss this with Jochen and Sebastian. If we move the attribute into local identifier then we should explicitly reference them in the KIT

same here

LuisRickert replied:
Yes the bpnlProerty must be the manufacturerID. We put it here because it is necessary to identify the related specification of the characteristic. Yes, the local identfier **may** contain the manufacturerId but it is not enforced by the datamodel. If we should use the localIdentifier for the manufacturerId this should be mentioned in the KIT and noted that this information is neccessary

johannsvarela commented on io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,63 @@
+:measurementType a samm:Property ;
+   samm:preferredName "Measurement Type"@en ;
+   samm:description "This optional attribute hints the type of the measurement"@en ;
+   samm:characteristic samm-c:Text .
```
Does it make sense to define this property as enumeration with the types, that will be defined in the KIT?

![image](https://github.com/user-attachments/assets/98f8aca6-03c4-4d4f-8221-2b944b6b88b4)

johannsvarela commented on io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,63 @@
+   samm:characteristic :CharacteristicCustomerPartID ;
+   samm:exampleValue "S1" .
+
+:results a samm:Property ;
```
The property  **results** is currently defined as a list of objects. However, the objects contained therein are not further defined. 

I'm against the definition of an “open” list of objects. Otherwise, there are no rules which objects are stored here, so that everything possible can be implemented and therefore the purpose of standardization becomes obsolete.

LuisRickert replied:
It was a explicit decision to keep it "open" because the set of characteristic is such a heterogenous set that one would have to create almost everytime a new object to fully capture the requirements for the measurement. I`ll raise to and discuss this with Jochen and Sebastian .

johannsvarela replied:
@LuisRickert I got your point of view and agree. But from a semantic perspective, it is **not recommended** to define an aspect model with an open object list. If we publish the model in this form, there won't be rules which data in which form are exchanged between business partners, as the data list is not further defined and therefore not standardized. This is **in contrary to the Catena-X approach**!

### Suggestion
As mentioned above, I recommend **defining all known and possible properties/attributes** for the object list for the version and expanding the aspect model with new data in the future as needed.

LuisRickert replied:
@johannsvarela after discussing this with Jochen and Sebastian we decided to keep the model as it is, since there is no sane way to provide a structure that could support a wide variety of measurements. We are aware that this can result in the rejection of the model.

Jochen-Fischer replied:
Hi @johannsvarela, @LuisRickert ,
there is one more important reason why we created the result as a set. The result is product specfic and not general usable for all product. Therefore no general aspect model for products in general possible.  There is also no traceability through multiple tiers possible, because every special characteristic is linked to one specific product.

johannsvarela reviewed (changes requested):
@LuisRickert Please have a look at my comments and change requests 😄 

@jSchuetz88 @markuskeidl FYI

johannsvarela commented on io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,82 @@
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
+@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
+@prefix : <urn:samm:io.catenax.specialCharacteristics.measurement:1.0.0#> .
```
> @LuisRickert for folder/path naming we don't use camel case but underscores. Therefore please change your new created folder _io.catenax.specia**lC**haracteristics.measurement/1.0.0/Measurement.ttl_ to _io.catenax.specia**l_c**haracteristics.measurement/1.0.0/Measurement.ttl_.

Same here... please replace the camel case with an underscore. 😄

LuisRickert replied:
done

LuisRickert commented:
> > ### Validation Report for io.catenax.specialCharacteristics.measurement/1.0.0/Measurement.ttl
> > #### Semantic violations were found:
> > ```turtle
> > > :Uuid :
> >   ---> Error at line 27 column 5
> >    | 
> > 27 |      samm:properties (:uuidV4Property ) ; ( ) ; ( :bpnaProperty ) ; . . ; ; . a  ; .
> >    |      ^^^^^^^^^^^^^^^ Property samm:properties is used 2 times on :Uuid, but may only be used 1 time.
> >    | 
> > 
> >   ---> Error at line 27 column 5
> >    | 
> > 27 |      samm:properties (:uuidV4Property ) ; ( ) ; ( :bpnaProperty ) ; . . ; ; . a  ; .
> >    |      ^^^^^^^^^^^^^^^ Property samm:properties is used 2 times on :Uuid, but may only be used 1 time.
> >    | 
> > ```
> 
> It appears that this is caused by the data model Serial Part. If the validation is run locally on the SerialPart the same error is returned.

Addressed by #790

LuisRickert commented:
Updated Example:
```json
{
  "localIdentifiers": [
    {
      "value": "SN12345678",
      "key": "partInstanceId"
    }
  ],
  "customerPartId": "PRT-12345",
  "bpnlProperty": "BPNL0123456789ZZ",
  "revisionIndex": "01",
  "measurementType": "simpleMeasurement",
  "characteristicId": "S1",
  "results": [
    {
      "identifier": "measurementPoint23",
      "description": "Lorem ipsum dolot sit amet, consetelur sadipscing eltir, sed diam nonumy eirmod tempr invidunt ut",
      "unit": "unit:hectopascal",
      "value": 12.09,
      "timestamp": "2024-11-18T14:52:42.867+01:00"
    }
  ]
}
```

johannsvarela commented on io.catenax.special_characteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,83 @@
+   samm:characteristic samm-c:Text ;
+   samm:exampleValue "measurementPoint23" .
+
+:timestamp a samm:Property ;
```
Don't use a generic name for this property. If you want to describe a measurement time, name it that way.

LuisRickert replied:
done

johannsvarela commented on io.catenax.special_characteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,83 @@
+   samm:characteristic :ValueCharacteristic ;
+   samm:exampleValue "12.09"^^xsd:float .
+
+:unit a samm:Property ;
```
Please use/reference the existing Quantity 2.0.0 aspect model to describe units: https://github.com/eclipse-tractusx/sldt-semantic-models/tree/main/io.catenax.shared.quantity/2.0.0

LuisRickert replied:
same questions as in the other comment

johannsvarela commented on io.catenax.special_characteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,83 @@
+   samm:description "Timestamp when the measurement was taken"@en ;
+   samm:characteristic samm-c:Timestamp .
+
+:value a samm:Property ;
```
Please use the existing Quantity 2.0.0 aspect model to describe units

LuisRickert replied:
I'm a bit confused that this model exists, what is the reasoning? It seems that this is just a wrapper around the predefined [units](https://eclipse-esmf.github.io/samm-specification/snapshot/appendix/unitcatalog.html). What would happen if there are units needed which are not represented in that model, e.g. Volt?

johannsvarela replied:
@LuisRickert interessting hint! @jSchuetz88 Lets discuss this topic at the semantic modelling meeting on monday afternoon!

LuisRickert replied:
As discussed in the meeting yesterday I won't apply this changes.

johannsvarela commented on io.catenax.special_characteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,83 @@
+   samm:preferredName "Description"@en ;
+   samm:description "Use this attribute to give a more detailed description to this measurement. "@en ;
+   samm:characteristic samm-c:Text ;
+   samm:exampleValue "Lorem ipsum dolot sit amet, consetelur sadipscing eltir, sed diam nonumy eirmod tempr invidunt ut" .
```
Please don't use a "lorem ipsum" example, since this will be a industry standard for special characteristics. The example should be useful to understand this attribute and therefore the whole model.

LuisRickert replied:
done

johannsvarela commented on io.catenax.special_characteristics.measurement/1.0.0/Measurement.ttl:
```diff
@@ -0,0 +1,83 @@
+:SpecialCharacteristicMeasurement a samm:Aspect ;
+   samm:preferredName "SpecialCharacteristicMeasurment"@en ;
+   samm:description "Contains the Measurment for a special characteristic"@en ;
+   samm:properties ( serialPart:localIdentifiers bpn:bpnlProperty serialPart:customerPartId :characteristicId :results [ samm:property :measurementType; samm:optional true ] :revisionIndex ) ;
```
To match the appropriate _SerialPart_ aspect model, use **manufacturerID** as name to describe the **bpnlProperty**. 

To do this, model a property called “manufactureID” and reference the _bpnlTrait_ from the BusinessPartnerNumber 2.0.0 aspect model, which contains the BPNL characteristics and constrains: https://github.com/eclipse-tractusx/sldt-semantic-models/tree/main/io.catenax.shared.business_partner_number/2.0.0

LuisRickert replied:
done

LuisRickert commented:
Updated Example:
Changes: 
- more specific attribute name
- no more lorem ipsum
- replaced bpn-property
```json
{
  "localIdentifiers": [
    {
      "value": "SN12345678",
      "key": "partInstanceId"
    }
  ],
  "customerPartId": "PRT-12345",
  "revisionIndex": "01",
  "manufacturerId": "BPN000000",
  "measurementType": "simpleMeasurement",
  "characteristicId": "S1",
  "results": [
    {
      "identifier": "measurementPoint23",
      "description": "This text gives you context to the related measurement to help to put it into the necessary context.",
      "unit": "unit:hectopascal",
      "value": 12.09,
      "measurementTimestamp": "2024-11-22T12:51:38.465+01:00"
    }
  ]
}
```

jSchuetz88 reviewed (commented):
Hi @LuisRickert,

thank you for the contribution (and your patience). So far I did not checked for the MS2 Criteria, as this PR needs some minor adjustments:

- [ ] The Aspect Model is named "SpecialCharacteristicMeasurement", but the file "Measurement". It needs to be the identitcal. I would suggest to rename the Measurement.ttl to "SpecialCharacteristicMeasurement.ttl"
- [ ] Please add a metadata.json file to the model with the content `{ "status" : "release"}`
- [x] Please add a RELEASE_NOTES.md file
- [x] Just for transparency, could you rename this PR to "[New Model]: SpecialCharacteristicMeasurement v.1.0.0"

jSchuetz88 reviewed (approved):
Looks good from my side. All MS II Criteria are met.

LuisRickert commented:
This should Not be closed, happend due to
